File: App/app/src/main/java/com/cinema/ticket_booking/util/ThemeManager.java
package com.cinema.ticket_booking.util;

import android.content.Context;
import android.content.SharedPreferences;
import androidx.appcompat.app.AppCompatDelegate;

/**
 * ThemeManager — quản lý Dark/Light mode
 * Lưu preference, apply khi khởi động app
 *
 * Dùng:
 * ThemeManager.applyTheme(context); // gọi trong Application.onCreate()
 * ThemeManager.toggleTheme(context); // gọi khi user bật/tắt switch
 * ThemeManager.isDarkMode(context); // kiểm tra mode hiện tại
 */
public class ThemeManager {

    private static final String PREFS_NAME = "nova_theme_prefs";
    private static final String KEY_DARK_MODE = "is_dark_mode";

    // Mặc định: theo system
    private static final int DEFAULT_MODE = AppCompatDelegate.MODE_NIGHT_FOLLOW_SYSTEM;

    /**
     * Apply theme đã lưu — gọi trong Application.onCreate() và
     * MainActivity.onCreate()
     */
    public static void applyTheme(Context context) {
        SharedPreferences prefs = getPrefs(context);
        boolean saved = prefs.contains(KEY_DARK_MODE);
        if (saved) {
            boolean isDark = prefs.getBoolean(KEY_DARK_MODE, false);
            AppCompatDelegate.setDefaultNightMode(
                    isDark ? AppCompatDelegate.MODE_NIGHT_YES : AppCompatDelegate.MODE_NIGHT_NO);
        } else {
            // Lần đầu: follow system
            AppCompatDelegate.setDefaultNightMode(DEFAULT_MODE);
        }
    }

    /** Toggle giữa dark và light, lưu preference */
    public static void toggleTheme(Context context) {
        boolean current = isDarkMode(context);
        setDarkMode(context, !current);
    }

    /** Set cụ thể dark hoặc light */
    public static void setDarkMode(Context context, boolean dark) {
        getPrefs(context).edit().putBoolean(KEY_DARK_MODE, dark).apply();
        AppCompatDelegate.setDefaultNightMode(
                dark ? AppCompatDelegate.MODE_NIGHT_YES : AppCompatDelegate.MODE_NIGHT_NO);
    }

    /** Kiểm tra mode hiện tại */
    public static boolean isDarkMode(Context context) {
        int currentMode = AppCompatDelegate.getDefaultNightMode();
        if (currentMode == AppCompatDelegate.MODE_NIGHT_YES)
            return true;
        if (currentMode == AppCompatDelegate.MODE_NIGHT_NO)
            return false;
        // Follow system — đọc từ resources
        int uiMode = context.getResources().getConfiguration().uiMode &
                android.content.res.Configuration.UI_MODE_NIGHT_MASK;
        return uiMode == android.content.res.Configuration.UI_MODE_NIGHT_YES;
    }

    private static SharedPreferences getPrefs(Context context) {
        return context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE);
    }
}

File: App/app/src/main/java/com/cinema/ticket_booking/util/ThemeManager.kt
package com.cinema.ticket_booking.util

import android.content.Context
import android.content.SharedPreferences
import android.content.res.Configuration
import androidx.appcompat.app.AppCompatDelegate

object ThemeManager {

    private const val PREFS_NAME = "nova_theme_prefs"
    private const val KEY_DARK_MODE = "is_dark_mode"

    private const val DEFAULT_MODE = AppCompatDelegate.MODE_NIGHT_FOLLOW_SYSTEM

    @JvmStatic
    fun applyTheme(context: Context) {
        val prefs = getPrefs(context)
        val saved = prefs.contains(KEY_DARK_MODE)
        if (saved) {
            val isDark = prefs.getBoolean(KEY_DARK_MODE, false)
            AppCompatDelegate.setDefaultNightMode(
                if (isDark) AppCompatDelegate.MODE_NIGHT_YES else AppCompatDelegate.MODE_NIGHT_NO
            )
        } else {
            AppCompatDelegate.setDefaultNightMode(DEFAULT_MODE)
        }
    }

    @JvmStatic
    fun toggleTheme(context: Context) {
        val current = isDarkMode(context)
        setDarkMode(context, !current)
    }

    @JvmStatic
    fun setDarkMode(context: Context, dark: Boolean) {
        getPrefs(context).edit().putBoolean(KEY_DARK_MODE, dark).apply()
        AppCompatDelegate.setDefaultNightMode(
            if (dark) AppCompatDelegate.MODE_NIGHT_YES else AppCompatDelegate.MODE_NIGHT_NO
        )
    }

    @JvmStatic
    fun isDarkMode(context: Context): Boolean {
        val currentMode = AppCompatDelegate.getDefaultNightMode()
        if (currentMode == AppCompatDelegate.MODE_NIGHT_YES) return true
        if (currentMode == AppCompatDelegate.MODE_NIGHT_NO) return false
        val uiMode = context.resources.configuration.uiMode and Configuration.UI_MODE_NIGHT_MASK
        return uiMode == Configuration.UI_MODE_NIGHT_YES
    }

    private fun getPrefs(context: Context): SharedPreferences {
        return context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
    }
}
